src/lib/draw/GeometryMode.svelte

```diff
@@ -38,88 +41,58 @@
-        delete feature.properties.hide_while_editing;
-        feature.geometry = editedRoute.geometry;
-        return gj;
-      });
 
-      // Stay in this mode
-      currentlyEditing = null;
-      currentlyEditingControls = null;
-    }
-  });
+        // Show the feature again
+        delete feature.properties.hide_while_editing;
 
-  routeTool.addEventListenerSuccessArea((editedArea) => {
-    if ($currentMode == thisMode) {
-      gjScheme.update((gj) => {
-        let feature = gj.features.find((f) => f.id == currentlyEditing)!;
-        if (!feature) {
-          window.alert(
-            "You loaded another file or cleared everything while editing. Your changes were lost."
-          );
-          return gj;
+        // If there are unsaved edits to the feature, copy them over. If the
+        // user explicitly canceled, then a failure callback would've run.
+        if (unsavedFeature) {
+          updateFeature(feature, unsavedFeature);
         }
-        // Keep the ID and any properties. Just copy over stuff from routeSnapper.
-        // TODO We're depending on implementation details here and knowing what to copy...
-        feature.properties.waypoints = editedArea.properties.waypoints;
-        delete feature.properties.hide_while_editing;
-        feature.geometry = editedArea.geometry;
         return gj;
       });
-
-      // Stay in this mode
-      currentlyEditing = null;
-      currentlyEditingControls = null;
     }
-  });
 
-  for (let tool of [pointTool, polygonTool]) {
-    tool.addEventListenerSuccess((feature) => {
+    stopEditing();
+    mapHover.set(null);
+  }
+
+  // Handle successful edits
+  for (let tool of [pointTool, polygonTool, routeTool]) {
+    tool.addEventListenerSuccess((editedFeature) => {
       if ($currentMode == thisMode) {
         gjScheme.update((gj) => {
-          let updateFeature = gj.features.find(
-            (f) => f.id == currentlyEditing
-          )!;
-          if (!updateFeature) {
+          let feature = gj.features.find((f) => f.id == currentlyEditing)!;
+          if (!feature) {
             window.alert(
               "You loaded another file or cleared everything while editing. Your changes were lost."
             );
             return gj;
           }
-          updateFeature.geometry = feature.geometry;
-          delete updateFeature.properties.hide_while_editing;
+          updateFeature(feature, editedFeature);
+          delete feature.properties.hide_while_editing;
           return gj;
         });
 
         // Stay in this mode
-        currentlyEditing = null;
-        currentlyEditingControls = null;
+        stopEditing();
+      }
+    });
+  }
+
+  // Auto-save for everything except pointTool
+  for (let tool of [polygonTool, routeTool]) {
+    tool.addEventListenerUpdated((feature) => {
+      if ($currentMode == thisMode) {
+        // Just remember the update; don't apply it yet
```

Comment:
We have a choice here. We could also save updates immediately as they happen, and to handle "Cancel", keep a copy of the feature before starting to edit. I've chosen not to do that, because saving when we switch mode is a bit more performant and is simpler. When we implement undo in #44, we can make the decision here to retain individual edits within a tool, or to compress them all like this. It's just a matter of doing something else in this callback.